os/drivers/error_report/error_report_drv.c

```diff
@@ -0,0 +1,301 @@
+ ****************************************************************************/
+#if CONFIG_ERROR_REPORT_INFINITE_WAIT
+#define INSERT_INFINITE_WAITQ(pos, tcb, cpu)        \
+		do {                                        \
+			g_thd_waitq[pos].pid = tcb->pid;        \
+			g_thd_waitq[pos].task_state = tcb->task_state;\
+			g_thd_waitq[pos].count = 1;             \
+			g_thd_waitq[pos].active_cpu = cpu.active;\
+			nwerr_vdbg("Inserted pid %d at pos %d\n", tcb->pid, pos);\
+		} while (0)
+
+#define DELETE_INFINITE_WAITQ(pos)                  \
+		do {                                        \
+			g_thd_waitq[pos].pid = 0;               \
+			g_thd_waitq[pos].count = 0;             \
+			g_thd_waitq[pos].task_state = 0;        \
+			g_thd_waitq[pos].active_cpu = 0;        \
+			nwerr_vdbg("Reset pos %d\n", pos);      \
+		} while (0)
+#endif
+
+static int prv_is_reserved(const char *taskname)
+{
+	int i;
+	if (taskname == NULL) {
+		return -1;
+	}
+	for (i = 0; i < sizeof(g_exceptional_waits)/sizeof(g_exceptional_waits[0]); i++) {
+		if (!strncmp(taskname, g_exceptional_waits[i], strlen(taskname))) {
+			return 1;
+		}
+	}
+	return 0;
+}
+
+static ssize_t err_report_read(FAR struct file *filep, FAR char *buffer, size_t len)
+{
+	return 0;
+}
+
+static ssize_t err_report_write(FAR struct file *filep, FAR const char *buffer, size_t len)
+{
+	return 0;
+}
+
+static unsigned long prv_fetch_taskaddr(int pid)
+{
+	struct tcb_s *tcbptr = sched_gettcb(pid);
+	if (tcbptr != NULL) {
+		entry_t e = tcbptr->entry;
+		if ((tcbptr->flags & TCB_FLAG_TTYPE_MASK) == TCB_FLAG_TTYPE_PTHREAD) {
+			return (unsigned long)e.pthread;
+		} else {
+			return (unsigned long)e.main;
+		}
+	}
+	return 0;
+}
+
+static void prv_tcb_handler(FAR struct tcb_s *tcb, void *args)
+{
+	int pos = 0;
+	if (tcb->pid < 5) {
+		return;
+	}
+	if ((tcb->task_state >= TSTATE_WAIT_SEM) && (tcb->task_state < NUM_TASK_STATES)) {
+		int ret;
+		struct cpuload_s cpu;
+		ret = clock_cpuload(tcb->pid, 0, &cpu);
+		if (ret != OK) {
+			return;
+		}       
+		/* Hash pid over the waitq range */
+		pos = tcb->pid % CONFIG_MAX_TASKS;
+		if ((g_thd_waitq[pos].pid == tcb->pid)) {
+			if (g_thd_waitq[pos].task_state == tcb->task_state) {
+				if (g_thd_waitq[pos].active_cpu == cpu.active) {
+					int threshold;
+					g_thd_waitq[pos].count++;
+					if (prv_is_reserved(tcb->name) == 1) {
+						threshold = INFINITE_CHECK_HIGH_THRESHOLD;
+					}
+					else {
+						threshold = CONFIG_ERROR_REPORT_INFINITE_CHECK_THRESHOLD;
+					}
+					
+					nwerr_vdbg("Inc wait count for pid %d\n", tcb->pid);
+					if (g_thd_waitq[pos].count >= threshold) {
+						// Add code for error report addition
+						uint32_t task_addr;
+						uint32_t saved_pc;
+						entry_t e = tcb->entry;
+						if ((tcb->flags & TCB_FLAG_TTYPE_MASK) == TCB_FLAG_TTYPE_PTHREAD) {
+							task_addr = (unsigned long)e.pthread;
+						} else {
+							task_addr = (unsigned long)e.main;
+						}					
+						saved_pc = tcb->xcp.regs[13];
```

Comment:
Is it enough to debug only with regs[13]?

Reply:
Thanks for the question. The tcb structure provides many fields to debug, of which regs[13] points to the stack. Assuming that developers would prefer something similar to a `gdb> backtrace`, we figured this would be a great place to start with. Please feel free to share any additional fields in tcb structure that can simplify debugging.